yetanotherco/aligned_layer
# fix(contracts): remove `ExceededMaxRespondFee` error

## Description

Removes `ExceededMaxRespondFee` error in `AlignedLayerServiceManager`:
- `ExceededMaxRespondFee` is removed from the Aligned contract and now the aggregator wallet will pay for the difference and the batcher will pay for the `feeLimit` he set. This case happens when the tx cost is higher than the `respondToTaskFeeLimit`.
- Metrics were added for: 
   - A count for the number of times the aggregator has paid for the difference.
   - An accumulated gas paid for the difference in eth.
  
<img width="651" alt="Screenshot 2024-11-13 at 3 30 29 PM" src="https://github.com/user-attachments/assets/1d9397e1-8f39-4405-b6bb-947d68663954">

- Now sending `respondToTask` in the aggregator will return a permanent error if and only if either the aggregator or the batcher doesn't have enough funds in their wallets.
 
 ## Testing


1. Modify the bumpedGasPrice percentage to 10000 so that the txCost is bigger than the `batch.RespondToTaskFeeLimit`:
```go
102: bumpedGasPrice := utils.CalculateGasPriceBumpBasedOnRetry(gasPrice, 10000, gasBumpIncrementalPercentage, I)
```
2. Setup a local testnet as usual.
3. Send a batch of proofs.
4. Check with the metrics and logs that the aggregator is paying for the batcher. 
5. Check the Batcher and Aggregator balance before and after the aggregator sends the response, the contract should've charged him with `batch.RespondToTaskFeeLimit`:
Get Agg balance before tx:
```shell
cast balance 0x15d34aaf54267db7d7c367839aaf71a00a2c6a65
```
Get batcher balance before transaction:
(Remember Batcher sends some funds when submitting task! so you must check the funds between the task submittion and the response)
```shell
cast call 0x851356ae760d987E095750cCeb3bC6014560891C "balanceOf(address)(uint256)" 0x7bc06c482dead17c0e297afbc32f6e63d3846650
```
  -  Get balances after tx.
  -  Aggregator should have decreased because he had to pay some of the tx:
  -  The difference in Batcher should be equal to `batch.RespondToTaskFeeLimit`. You can identify the `RespondToTaskFeeLimit` value from the aggregator logs.


Another test:
- you can undo the changes in the contract, re-deploy contracts in anvil, and check that the `RespondToTask` reverts.


## Type of change



## Checklist

  - [Link to the open issue addressing it](#1402 )
